sched's Travis build fails after pmix update

Travis build for sched fails, probably because libevent-dev (required by pmix) is missing. 

```
============================================================================
== Libevent
============================================================================
checking for libevent in... (default search paths)
looking for header in /usr/include
checking event.h usability... no
checking event.h presence... no
checking for event.h... no
looking for header in /usr/include/include
checking event.h usability... no
checking event.h presence... no
checking for event.h... no
configure: WARNING: LIBEVENT SUPPORT NOT FOUND
configure: error: CANNOT CONTINUE
make: *** No targets specified and no makefile found.  Stop.
travis-dep-builder.sh: Failed to build and install pmix-3.0.0
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -51,6 +51,7 @@ addons:
       - libboost-all-dev
       - libreadline6
       - valgrind
+      - libevent-dev
 
 before_install:
   - wget https://raw.githubusercontent.com/flux-framework/flux-core/master/src/test/travis-dep-builder.sh

--- a/sched/sched.c
+++ b/sched/sched.c
@@ -74,7 +74,12 @@ static void exclude_request_cb (flux_t *h, flux_msg_handler_t *w,
                               const flux_msg_t *msg, void *arg);
 static void include_request_cb (flux_t *h, flux_msg_handler_t *w,
                               const flux_msg_t *msg, void *arg);
+static void sched_params_set_request_cb (flux_t *h, flux_msg_handler_t *w,
+                              const flux_msg_t *msg, void *arg);
+static void sched_params_get_request_cb (flux_t *h, flux_msg_handler_t *w,
+                              const flux_msg_t *msg, void *arg);
 static int job_status_cb (const char *jcbstr, void *arg, int errnum);
+
 
 
 /******************************************************************************
@@ -139,6 +144,7 @@ typedef struct {
     flux_msg_handler_t **sim_handlers;
 } ssrvctx_t;
 
+static int schedule_jobs (ssrvctx_t *ctx);  /* Forward declaration */
 
 /******************************************************************************
  *                                                                            *
@@ -305,28 +311,70 @@ static int adjust_for_sched_params (ssrvctx_t *ctx)
 {
     flux_reactor_t *r = NULL;
     int rc = 0;
+    flux_msg_t *msg = NULL; 
+    /* delay_sched = true. 
+    Watchers need to be created the first time delay_sched is set to true. 
+    Otherwise, start the prepare/check watchers. The idle watcher is started by the ev_prep_cb. */ 
     if (ctx->arg.s_params.delay_sched && !ctx->sctx.in_sim) {
-        if (!(r = flux_get_reactor (ctx->h))) {
-            rc = -1;
-            goto done;
+        if (ctx->before && ctx->after) {                    /* Watchers exist */ 
+            flux_watcher_start (ctx->before); 
+            flux_watcher_start (ctx->after);
         }
-        if (!(ctx->before = flux_prepare_watcher_create (r, ev_prep_cb, ctx))) {
-            rc = -1;
-            goto done;
+       
+        /* If one of the watchers is not available, send error */
+        if ((!ctx->before && ctx->after) || (!ctx->after && ctx->before)) {
+            rc = -1; 
+            errno = EINVAL;
+            goto done; 
+        } 
+ 
+        if (!ctx->before && !ctx->after) {      /* Create and start watchers */
+            if (!(r = flux_get_reactor (ctx->h))) {
+                rc = -1;
+                goto done;
+            }
+            if (!(ctx->before = flux_prepare_watcher_create (r, ev_prep_cb, ctx))) {
+                rc = -1;
+                goto done;
+            }
+            if (!(ctx->after = flux_check_watcher_create (r, ev_check_cb, ctx))) {
+                rc = -1;
+                goto done;
+            }
+            /* idle watcher makes sure the check watcher (after) is called
+            even with no external events delivered */
+            if (!(ctx->idle = flux_idle_watcher_create (r, NULL, NULL))) {
+                rc = -1;
+                goto done;
+            }
+            flux_watcher_start (ctx->before);
+            flux_watcher_start (ctx->after);
         }
-        if (!(ctx->after = flux_check_watcher_create (r, ev_check_cb, ctx))) {
-            rc = -1;
-            goto done;
+    } // End if, delay = true
+
+    /* delay_sched = false. 
+    If set at runtime, stop the watchers, call schedule_jobs */
+    if (!ctx->arg.s_params.delay_sched && !ctx->sctx.in_sim) {
+        if (ctx->before && ctx->after) {
+            flux_watcher_stop (ctx->before); 
+            flux_watcher_stop (ctx->after);
+        
+            /* Create an event to schedule_jobs */   
+            flux_log (ctx->h, LOG_DEBUG, "Update delay_sched parameter");
+            msg = flux_event_encode ("sched.res.param_update", NULL);
+            if (!msg || flux_send (ctx->h, msg, 0) < 0) {
+                flux_log (ctx->h, LOG_ERR, "%s: error sending event: %s",
+                __FUNCTION__, strerror (errno));
+            }
+            flux_msg_destroy (msg);
         }
-        /* idle watcher makes sure the check watcher (after) is called
-           even with no external events delivered */
-        if (!(ctx->idle = flux_idle_watcher_create (r, NULL, NULL))) {
+        /* If one of the watchers is not available, send error */
+        if ((!ctx->before && ctx->after) || (!ctx->after && ctx->before)) {
             rc = -1;
-            goto done;
-        }
-        flux_watcher_start (ctx->before);
-        flux_watcher_start (ctx->after);
-    }
+            errno = EINVAL;
+            goto done; 
+        } 
+    } // End if, delay = false 
 done:
     return rc;
 }
@@ -1077,6 +1125,8 @@ static const struct flux_msg_handler_spec htab[] = {
     { FLUX_MSGTYPE_REQUEST,   "sched.cancel", cancel_request_cb, 0},
     { FLUX_MSGTYPE_REQUEST,   "sched.exclude",  exclude_request_cb, 0},
     { FLUX_MSGTYPE_REQUEST,   "sched.include",  include_request_cb, 0},
+    { FLUX_MSGTYPE_REQUEST,   "sched.params.set", sched_params_set_request_cb, 0},
+    { FLUX_MSGTYPE_REQUEST,   "sched.params.get", sched_params_get_request_cb, 0},
     { FLUX_MSGTYPE_EVENT,     "sched.res.*",  res_event_cb, 0},
     FLUX_MSGHANDLER_TABLE_END
 };
@@ -2084,6 +2134,70 @@ static int job_status_cb (const char *jcbstr, void *arg, int errnum)
 out:
     Jput (jcb);
     return rc;
+}
+
+
+static void sched_params_set_request_cb (flux_t *h, flux_msg_handler_t *w,
+                                const flux_msg_t *msg, void *arg)
+{
+    ssrvctx_t *ctx = getctx ((flux_t *)arg);
+    uint32_t userid = 0;
+    char *sprms = NULL;
+    bool prev_delay_sched = ctx->arg.s_params.delay_sched;
+
+    if (flux_msg_get_userid (msg, &userid) < 0)
+        goto error;
+
+    flux_log (h, LOG_INFO, 
+            "sched_params change requested by user (%u).", userid);
+    
+    if (flux_request_unpack (msg, NULL, "{s:s}", "param", &sprms) < 0)
+        goto error;
+   
+    if (sched_params_args (sprms, &(ctx->arg.s_params)) != 0)
+        goto error;
+   
+    /* Only call adjust_for_sched_params if there's a state change in delay_sched, 
+    otherwise, flux_watcher_start gets called more than once */
+    if (prev_delay_sched != ctx->arg.s_params.delay_sched) {
+        if (adjust_for_sched_params (ctx) != 0) 
+            goto error;
+    }
+   
+    if (flux_respond_pack (h, msg, "{}") < 0)
+        flux_log_error (h, "%s", __FUNCTION__);
+    return;
+
+error:
+    if (flux_respond (h, msg, errno, NULL) < 0)
+        flux_log_error (h, "%s", __FUNCTION__);
+}
+
+
+static void sched_params_get_request_cb (flux_t *h, flux_msg_handler_t *w,
+                                const flux_msg_t *msg, void *arg)
+{
+    ssrvctx_t *ctx = getctx ((flux_t *)arg);
+    uint32_t userid = 0;
+
+    if (flux_msg_get_userid (msg, &userid) < 0)
+        goto error;
+
+    flux_log (h, LOG_INFO, 
+            "sched_param values requested by user (%u).", userid);
+   
+     if (flux_request_unpack (msg, NULL, "{}") < 0)
+        goto error;
+
+    if (flux_respond_pack (h, msg, "{s:i s:i}", 
+            "queue-depth", (int)ctx->arg.s_params.queue_depth, 
+            "delay-sched", (int)ctx->arg.s_params.delay_sched) < 0)
+        flux_log_error (h, "%s", __FUNCTION__);
+    return;
+
+error:
+    if (flux_respond (h, msg, errno, NULL) < 0)
+        flux_log_error (h, "%s", __FUNCTION__);
 }
 
 static void ev_prep_cb (flux_reactor_t *r, flux_watcher_t *w, int ev, void *a)

--- a/t/Makefile.am
+++ b/t/Makefile.am
@@ -48,6 +48,7 @@ TESTS = \
     t1003-stress.t \
     t1004-module-load.t \
     t1005-sched-params.t \
+    t1008-runtime-sched-params.t \
     t1006-cancel.t \
     t1007-exclude.t \
     t2000-fcfs.t \

--- a/t/t1008-runtime-sched-params.t
+++ b/t/t1008-runtime-sched-params.t
@@ -1,0 +1,186 @@
+#!/bin/bash
+set -x
+
+test_description='Test dynamic updation of queue depth. 
+Ensure jobs are correctly scheduled under different values of
+scheduling optimization parameters and their combinations
+'
+
+. `dirname $0`/sharness.sh
+
+basepath=`readlink -e ${SHARNESS_TEST_SRCDIR}/data/hwloc-data`
+# each of the 4 brokers manages a full cab node exclusively
+excl_4N4B=$basepath/004N/exclusive/04-brokers
+excl_4N4B_nc=16
+
+#
+# test_under_flux is under sharness.d/
+#
+test_under_flux 4
+
+#
+# print only with --debug
+#
+test_debug '
+    echo ${basepath} &&
+    echo ${excl_4N4B} &&
+    echo ${excl_4N4B_nc}
+'
+
+ test_expect_success 'sched-params: getting params (no args) at runtime works' '
+    adjust_session_info 4 &&
+    flux hwloc reload ${excl_4N4B} &&
+    flux module load sched  sched-once=true sched-params=queue-depth=1024,delay-sched=true &&
+    timed_wait_job 5 &&
+    submit_1N_nproc_sleep_jobs ${excl_4N4B_nc} 0 &&
+    flux wreck sched-params get > output &&
+    printf "delay-sched=true\nqueue-depth=1024\n" > expected &&
+    test_cmp output expected &&  
+    timed_sync_wait_job 10 &&
+    verify_1N_nproc_sleep_jobs ${excl_4N4B_nc} 
+'
+ test_expect_success 'sched-params: getting queue-depth at runtime works' '
+    adjust_session_info 4 &&
+    flux module remove sched &&
+    flux hwloc reload ${excl_4N4B} &&
+    flux module load sched  sched-once=true sched-params=queue-depth=1024,delay-sched=true &&
+    timed_wait_job 5 &&
+    submit_1N_nproc_sleep_jobs ${excl_4N4B_nc} 0 &&
+    flux wreck sched-params get queue-depth > output &&
+    printf "queue-depth=1024\n" > expected &&
+    test_cmp output expected &&  
+    timed_sync_wait_job 10 &&
+    verify_1N_nproc_sleep_jobs ${excl_4N4B_nc} 
+' 
+ test_expect_success 'sched-params: getting delay-sched at runtime works' '
+    adjust_session_info 4 &&
+    flux module remove sched &&
+    flux hwloc reload ${excl_4N4B} &&
+    flux module load sched  sched-once=true sched-params=queue-depth=1024,delay-sched=true &&
+    timed_wait_job 5 &&
+    submit_1N_nproc_sleep_jobs ${excl_4N4B_nc} 0 &&
+    flux wreck sched-params get delay-sched > output &&
+    printf "delay-sched=true\n" > expected &&
+    test_cmp output expected &&  
+    timed_sync_wait_job 10 &&
+    verify_1N_nproc_sleep_jobs ${excl_4N4B_nc} 
+'
+test_expect_success 'sched-params: getting both params at runtime works' '
+    adjust_session_info 4 &&
+    flux module remove sched &&
+    flux hwloc reload ${excl_4N4B} &&
+    flux module load sched sched-once=true sched-params=queue-depth=1024,delay-sched=false &&
+    timed_wait_job 5 &&
+    submit_1N_nproc_sleep_jobs ${excl_4N4B_nc} 0 &&
+    flux wreck sched-params get queue-depth,delay-sched > output &&
+    printf "delay-sched=false\nqueue-depth=1024\n" > expected &&
+    test_cmp output expected &&  
+    timed_sync_wait_job 10 &&
+    verify_1N_nproc_sleep_jobs ${excl_4N4B_nc} 
+ '
+test_expect_success 'sched-params: update/get works with a short queue depth (4)' '
+    adjust_session_info 4 &&
+    flux module remove sched &&
+    flux hwloc reload ${excl_4N4B} &&
+    flux module load sched sched-once=true sched-params=queue-depth=1024 &&
+    timed_wait_job 5 &&
+    submit_1N_nproc_sleep_jobs ${excl_4N4B_nc} 0 &&
+    test `flux wreck sched-params get queue-depth | cut -d "=" -f 2` -eq 1024 &&
+    flux wreck sched-params set queue-depth=4 &&
+    test `flux wreck sched-params get queue-depth | cut -d "=" -f 2` -eq 4 &&
+    timed_sync_wait_job 10 &&
+    verify_1N_nproc_sleep_jobs ${excl_4N4B_nc} 
+'
+ test_expect_success 'sched-params: update/get works with another queue-depth (16)' '
+    adjust_session_info 4 &&
+    flux module remove sched &&
+    flux hwloc reload ${excl_4N4B} &&
+    flux module load sched sched-once=true sched-params=queue-depth=4 &&
+    timed_wait_job 5 &&
+    submit_1N_nproc_sleep_jobs ${excl_4N4B_nc} 0 &&
+    test `flux wreck sched-params get queue-depth | cut -d "=" -f 2` -eq 4 &&
+    flux wreck sched-params set queue-depth=16 &&
+    test `flux wreck sched-params get queue-depth | cut -d "=" -f 2` -eq 16 &&
+    timed_sync_wait_job 10 &&
+    verify_1N_nproc_sleep_jobs ${excl_4N4B_nc} 
+ '
+test_expect_success 'sched-params: update/get works with a long queue depth (4096)' '
+    adjust_session_info 4 &&
+    flux module remove sched &&
+    flux hwloc reload ${excl_4N4B} &&
+    flux module load sched sched-once=true sched-params=queue-depth=2048 &&
+    timed_wait_job 5 &&
+    submit_1N_nproc_sleep_jobs ${excl_4N4B_nc} 0 &&
+    test `flux wreck sched-params get queue-depth | cut -d "=" -f 2` -eq 2048 &&
+    flux wreck sched-params set queue-depth=4096 &&
+    test `flux wreck sched-params get queue-depth | cut -d "=" -f 2` -eq 4096 &&
+    timed_sync_wait_job 10 &&
+    verify_1N_nproc_sleep_jobs ${excl_4N4B_nc} 
+'
+test_expect_success 'sched-params: update/get works with several delay_sched changes' '
+    adjust_session_info 4 &&
+    flux module remove sched &&
+    flux hwloc reload ${excl_4N4B} &&
+    flux module load sched sched-once=true sched-params=delay-sched=true &&
+    timed_wait_job 5 &&
+    submit_1N_nproc_sleep_jobs ${excl_4N4B_nc} 0 &&
+    test `flux wreck sched-params get delay-sched | cut -d "=" -f 2` &&
+    flux wreck sched-params set delay-sched=false &&
+    test !`flux wreck sched-params get delay-sched | cut -d "=" -f 2` &&
+    flux wreck sched-params set delay-sched=true &&
+    test `flux wreck sched-params get delay-sched | cut -d "=" -f 2` &&
+    flux wreck sched-params set delay-sched=false &&
+    test !`flux wreck sched-params get delay-sched | cut -d "=" -f 2` &&
+    flux wreck sched-params set delay-sched=true &&
+    test !`flux wreck sched-params get delay-sched | cut -d "=" -f 2` &&
+    timed_sync_wait_job 10 &&
+    verify_1N_nproc_sleep_jobs ${excl_4N4B_nc} 
+'
+test_expect_success 'sched-params: update/get works with delay-sched=true' '
+    adjust_session_info 4 &&
+    flux module remove sched &&
+    flux hwloc reload ${excl_4N4B} &&
+    flux module load sched sched-once=true sched-params=delay-sched=true &&
+    timed_wait_job 5 &&
+    submit_1N_nproc_sleep_jobs ${excl_4N4B_nc} 0 &&
+    test `flux wreck sched-params get delay-sched | cut -d "=" -f 2` &&
+    flux wreck sched-params set delay-sched=true &&
+    test `flux wreck sched-params get delay-sched | cut -d "=" -f 2` &&
+    timed_sync_wait_job 10 &&
+    verify_1N_nproc_sleep_jobs ${excl_4N4B_nc} 
+'
+test_expect_success 'sched-params: update to delay can be combined with a short depth' '
+    adjust_session_info 4 &&
+    flux module remove sched &&
+    flux hwloc reload ${excl_4N4B} &&
+    printf  "delay-sched=true\nqueue-depth=16\n" > expected1 &&
+    printf "delay-sched=true\nqueue-depth=32\n" > expected2 &&
+    flux module load sched sched-once=true \
+    sched-params=delay-sched=true,queue-depth=16 &&
+    timed_wait_job 5 &&
+    submit_1N_nproc_sleep_jobs ${excl_4N4B_nc} 0 &&
+    flux wreck sched-params get > output1 &&
+    flux wreck sched-params set queue-depth=32,delay-sched=true &&
+    flux wreck sched-params get > output2 &&
+    test_cmp output1 expected1 &&  
+    test_cmp output2 expected2 &&  
+    timed_sync_wait_job 10 &&
+    verify_1N_nproc_sleep_jobs ${excl_4N4B_nc} 
+'
+test_expect_success 'sched-params: update to delay can be made individually' '
+    adjust_session_info 4 &&
+    flux module remove sched &&
+    flux hwloc reload ${excl_4N4B} &&
+    flux module load sched sched-once=true \
+    sched-params=queue-depth=2048 &&
+    timed_wait_job 5 &&
+    submit_1N_nproc_sleep_jobs ${excl_4N4B_nc} 0 &&
+    flux wreck sched-params set delay-sched=false &&
+    test `flux wreck sched-params get delay-sched | cut -d "=" -f 2` &&
+    timed_sync_wait_job 10 &&
+    verify_1N_nproc_sleep_jobs ${excl_4N4B_nc} 
+'
+test_expect_success 'sched-params: unloaded sched module' '
+    flux module remove sched
+'
+test_done

